### PR TITLE
chore: release v0.21

### PR DESCRIPTION
**Pull request by niklasad1**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,137 @@ The format is based on [Keep a Changelog].
 
 [Keep a Changelog]: http://keepachangelog.com/en/1.0.0/
 
+## [v0.21.0] - 2023-12-11
+
+This is release contains big changes and let's go over the main ones:
```
Comment (lexnv) [inline CHANGELOG.md:9]:
```suggestion
This release contains big changes and let's go over the main ones:
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,137 @@ The format is based on [Keep a Changelog].
...
+
+### JSON-RPC specific middleware
+
+After getting plenty of feedback from regarding a JSON-RPC specific middleware,
```
Comment (lexnv) [inline CHANGELOG.md:13]:
```suggestion
After getting plenty of feedback regarding a JSON-RPC specific middleware,
```

**Review by lexnv — APPROVED** [review]
This is a huge release indeed! Nice one! 👍

**Review by jsdw — APPROVED** [review]
Awesome; will be great to get this out!

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,137 @@ The format is based on [Keep a Changelog].
...
+
+### jsonrpsee server low-level API
+
+For users who wants to get low-level access and for example disconnect
+misbehaving peers that is now possible as well [example here](./examples/examples/jsonrpsee_server_low_level_api.rs)
+
+### Logging in the server
+
+Logging of RPC calls has been disabled by default, 
+but it's possible to enable that with the RPC middleware.
```
Comment (niklasad1) [inline CHANGELOG.md:62]:
```suggestion
but it's possible to enable that with the RPC logger middleware.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,137 @@ The format is based on [Keep a Changelog].
...
+		}
+		// Re-direct all calls that isn't `say_hello` to `say_goodbye`
+		else if req.method != "say_hello" {
+			req.method = "say_goodbye".into();
+		}
+
+		self.0.call(req)
+	}
+}
+```
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply CHANGELOG.md:47]: ok, yeah I see your point now :P

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,152 @@ The format is based on [Keep a Changelog].
...
+async fn run_server() {
+	// Construct our middleware and build the server.
+	let rpc_middleware = RpcServiceBuilder::new().layer_fn(|service| ModifyRequestIf(service));
+	let server = Server::builder().set_rpc_middleware(rpc_middleware).build("127.0.0.1:0").await.unwrap();
+
+	// Start the server.
+	let mut module = RpcModule::new(());
+	module.register_method("say_hello", |_, _| "lo").unwrap();
+	module.register_method("say_goodbye", |_, _| "goodbye").unwrap();
+	let addr = server.local_addr()?;
```
Comment (niklasad1) [inline CHANGELOG.md:57]:
```suggestion
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,151 @@ The format is based on [Keep a Changelog].
...
+- chore(deps): update tokio-rustls requirement from 0.24 to 0.25 ([#1256](https://github.com/paritytech/jsonrpsee/pull/1256))
+- chore(deps): update gloo-net requirement from 0.4.0 to 0.5.0 ([#1260](https://github.com/paritytech/jsonrpsee/pull/1260))
+- chore(deps): update async-lock requirement from 2.4 to 3.0  ([#1226](https://github.com/paritytech/jsonrpsee/pull/1226))
+- chore(deps): update proc-macro-crate requirement from 1 to 2  ([#1211](https://github.com/paritytech/jsonrpsee/pull/1211))
+- chore(deps): update console-subscriber requirement from 0.1.8 to 0.2.0  ([#1210](https://github.com/paritytech/jsonrpsee/pull/1210))
+- refactor: split client and server errors ([#1122](https://github.com/paritytech/jsonrpsee/pull/1122))
+- refactor(ws client): impl tokio:{AsyncRead, AsyncWrite} for EitherStream ([#1249](https://github.com/paritytech/jsonrpsee/pull/1249))
+- http client: enable all http versions ([#1252](https://github.com/paritytech/jsonrpsee/pull/1252))
+- server: change ws ping API  ([#1248](https://github.com/paritytech/jsonrpsee/pull/1248))
+- refactor: update and turn `WsTransportClientBuilder` generic ([#1168](https://github.com/paritytech/jsonrpsee/pull/1168))
```
Comment (niklasad1) [inline CHANGELOG.md:101]:
```suggestion
- refactor(client): WS client generic over data stream  ([#1168](https://github.com/paritytech/jsonrpsee/pull/1168))
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,151 @@ The format is based on [Keep a Changelog].
...
+### [Changed]
+- chore(deps): update tokio-rustls requirement from 0.24 to 0.25 ([#1256](https://github.com/paritytech/jsonrpsee/pull/1256))
+- chore(deps): update gloo-net requirement from 0.4.0 to 0.5.0 ([#1260](https://github.com/paritytech/jsonrpsee/pull/1260))
+- chore(deps): update async-lock requirement from 2.4 to 3.0  ([#1226](https://github.com/paritytech/jsonrpsee/pull/1226))
+- chore(deps): update proc-macro-crate requirement from 1 to 2  ([#1211](https://github.com/paritytech/jsonrpsee/pull/1211))
+- chore(deps): update console-subscriber requirement from 0.1.8 to 0.2.0  ([#1210](https://github.com/paritytech/jsonrpsee/pull/1210))
+- refactor: split client and server errors ([#1122](https://github.com/paritytech/jsonrpsee/pull/1122))
+- refactor(ws client): impl tokio:{AsyncRead, AsyncWrite} for EitherStream ([#1249](https://github.com/paritytech/jsonrpsee/pull/1249))
+- refactor(http client): enable all http versions ([#1252](https://github.com/paritytech/jsonrpsee/pull/1252))
+- server: change ws ping API  ([#1248](https://github.com/paritytech/jsonrpsee/pull/1248))
```
Comment (niklasad1) [inline CHANGELOG.md:100]:
```suggestion
- refactor(server): change ws ping API  ([#1248](https://github.com/paritytech/jsonrpsee/pull/1248))
```

**Comment by niklasad1** [comment]
We'll publish this release next week because I want to do some manual tests before publishing

**Pull request merged**